crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+    ///
+    /// If this is the result of a forward pagination, then the events are in
+    /// topological order.
+    pub events: Vec<TimelineEvent>,
+
+    /// Did we hit an end of the timeline?
+    ///
+    /// If this is the result of a backward pagination, this means we hit the
+    /// *start* of the timeline.
+    ///
+    /// If this is the result of a forward pagination, this means we hit the
+    /// *end* of the timeline.
+    pub hit_end_of_timeline: bool,
+}
+
+/// The result of an initial [`Paginator::start_from`] query.
+#[derive(Debug)]
+pub struct StartFromResult {
+    /// All the events returned during this pagination, in topological ordering.
+    pub events: Vec<TimelineEvent>,
+
+    /// Whether the /context query returned a previous batch token.
+    pub has_prev: bool,
+
+    /// Whether the /context query returned a next batch token.
+    pub has_next: bool,
+}
+
+impl Paginator {
+    /// Create a new [`Paginator`], given a room implementation.
+    pub fn new(room: Box<dyn PaginableRoom>) -> Self {
+        Self {
+            room,
+            state: SharedObservable::new(PaginatorState::Initial),
+            prev_batch_token: Mutex::new(None),
+            next_batch_token: Mutex::new(None),
+        }
+    }
+
+    /// Check if the current state of the paginator matches the expected one.
+    fn check_state(&self, expected: PaginatorState) -> Result<(), PaginatorError> {
+        let actual = self.state.get();
+        if actual != expected {
+            return Err(PaginatorError::InvalidPreviousState { expected, actual });
+        }
+        Ok(())
+    }
+
+    /// Returns a subscriber to the internal [`PaginatorState`] machine.
+    pub fn state(&self) -> Subscriber<PaginatorState> {
+        self.state.subscribe()
+    }
+
+    /// Starts the pagination from the initial event.
+    ///
+    /// Only works for fresh [`Paginator`] objects, which are in the
+    /// [`PaginatorState::Initial`] state.
+    pub async fn start_from(&self, event_id: &EventId) -> Result<StartFromResult, PaginatorError> {
+        self.check_state(PaginatorState::Initial)?;
+
+        // Note: it's possible two callers have checked the state and both figured it's
+        // initial. This check makes sure there's at most one which can set the
+        // state to FetchingTargetEvent, preventing a race condition.
+        if self.state.set_if_not_eq(PaginatorState::FetchingTargetEvent).is_none() {
+            return Err(PaginatorError::InvalidPreviousState {
+                expected: PaginatorState::Initial,
+                actual: PaginatorState::FetchingTargetEvent,
+            });
+        }
+
+        // TODO: do we want to lazy load members?
+        let lazy_load_members = true;
+
+        let response = self.room.event_with_context(event_id, lazy_load_members).await?;
+
+        let has_prev = response.prev_batch_token.is_some();
+        let has_next = response.next_batch_token.is_some();
```

Comment:
Maybe move these 2 `is_some` lines down to just before these variables are used?

Reply:
Oh, no, I see! You are about to change them.

Reply:
Can we name them somehow to make clear these are the old values?

Reply:
Ah, I was confused.

Then I go back to my original comment: why not create these variables immediately before they are used instead of creating them a few lines earlier? (Very minor point)

Reply:
Because that would require taking the lock again, since the `response`'s ownership is partially taken when assigning the inner fields on the next two lines.